src/app/CommandSender.h

```diff
@@ -436,8 +516,59 @@ class CommandSender final : public Messaging::ExchangeDelegate
 
     CHIP_ERROR SendCommandRequestInternal(const SessionHandle & session, Optional<System::Clock::Timeout> timeout);
 
+    void OnResponseCallback(const ConcreteCommandPath & aPath, const StatusIB & aStatusIB, TLV::TLVReader * apData,
+                            const AdditionalResponseData & aAdditionalResponseData)
+    {
+        if (mUsingExtendedCallbacks)
+        {
+            if (mpExtendedCallback)
+            {
+                mpExtendedCallback->OnResponse(this, aPath, aStatusIB, apData, aAdditionalResponseData);
+            }
+            return;
+        }
+        if (mpCallback)
+        {
+            mpCallback->OnResponse(this, aPath, aStatusIB, apData);
+        }
+    }
+
+    void OnErrorCallback(CHIP_ERROR aError)
+    {
+        if (mUsingExtendedCallbacks)
+        {
+            if (mpExtendedCallback)
+            {
+                ErrorData errorData = { aError };
+                mpExtendedCallback->OnError(this, errorData);
+            }
+            return;
+        }
+        if (mpCallback)
+        {
+            mpCallback->OnError(this, aError);
+        }
+    }
+
+    void OnDoneCallback()
+    {
+        if (mUsingExtendedCallbacks)
+        {
+            if (mpExtendedCallback)
+            {
+                mpExtendedCallback->OnDone(this);
+            }
+            return;
```

Comment:
I am not sure the early return is worth it if mpCallback and mpExtendedCallback are always mutually exclusive. We could simply have:

```
if (mpExtendedCallback) 
{
    mpExtendedCallback->OnDone(this);
}
if (mpCallback)
{
    mpCallback->OnDone(this);
}
```

And no early return, and no `if (mUsingExtendedCallbacks)`. In fact, with the new explicit constructors, `mUsingExtendedCallbacks` is implicitly `mpExtendedCallback != nullptr` and the bool for `mUsingExtendedCallbacks` can be removed entirely from the class.

Reply:
we should go with early return because mpCallback and mpExtendedCallback are always mutually exclusive.
If they both have been set, we should return error, maybe check in PrepareCommand.

Reply:
Check added in `AllocateBuffer`